File: PythonCode/marg_csmc.py
######################################################################################
#
# Author : Niharika Gauraha
#          KTH
#          Email : niharika@kth.se
#
# Published under GNU General Public License
#####################################################################################

# Implements Marginalized Conditional SMC (MCSMC)

import numpy as np
from utils import multinomial_resampling
import sys
import scipy.stats as stats


class MargCSMC(object):
    def __init__(self, f, g, prior):
        if (f is None) or (g is None):
            sys.exit("Error: the input parameters are NULL")
        self._f = f
        self._g = g  # current global model
        self._resampling_method = multinomial_resampling
        self._particles = None
        self._ancestors = None
        self._normalised_weights = None
        self._seq_len = 0
        self._n_particles = 100
        self._a0 = prior[0]
        self._b0 = prior[1]
        self._c0 = prior[2]
        self._d0 = prior[3]
        self._alpha_x = 0.0
        self._alpha_y = 0.0
        self._beta_x = None
        self._beta_y = None
        self._sx = None
        self._sy = None

    def _resample_step(self, t):
        new_ancestors = self._resampling_method(self._normalised_weights[:, t])
        self._ancestors[:, t] = new_ancestors.astype(int)

    def _propagation_step(self, t):
        xpred = self._f(self._particles[:, t - 1], t - 1)
        nu = 2 * self._alpha_x
        sig2 = self._beta_x / self._alpha_x
        ancestors = self._ancestors[:, t - 1]
        self._particles[:, t] = xpred[ancestors] + np.sqrt(sig2) * stats.t.rvs(df=nu, size=self._n_particles)

    def _weighting_step(self, y_t, t):
        if t == 0:
            nu = 2 * self._alpha_y
            sig2 = self._beta_y / self._alpha_y
            ypred = self._g(self._particles[:, t])
            weights = stats.t.logpdf(y_t, df=nu, loc=ypred, scale=np.sqrt(sig2))

            # compute normalized weights
            max_weight = max(weights)  # Subtract the maximum value for numerical stability
            w = np.exp(weights - max_weight)
            self._normalised_weights[:, t] = w / sum(w)

            # sx = sx[newAncestors] + -0.5 * (particles[:, t] - xpred[newAncestors]) ** 2
            self._sy = - 0.5 * (y_t - ypred) ** 2
        else:
            nu = 2 * self._alpha_y
            sig2 = self._beta_y / self._alpha_y
            xpred = self._f(self._particles[:, t-1], t-1)
            ypred = self._g(self._particles[:, t])
            weights =  stats.t.logpdf(y_t, df=nu, loc=ypred, scale=np.sqrt(sig2))
            max_weight = max(weights)  # Subtract the maximum value for numerical stability
            w = np.exp(weights - max_weight)
            self._normalised_weights[:, t] = w / sum(w)
            ancestors = self._ancestors[:, t-1]
            self._sx = self._sx[ancestors] - 0.5 * (self._particles[:, t] - xpred[ancestors]) ** 2
            self._sy = self._sy[ancestors] - 0.5 * (y_t - ypred) ** 2

        self._alpha_x = self._alpha_x + 1 / 2
        self._beta_x = self._b0 - self._sx
        self._alpha_y = self._alpha_y + 1 / 2
        self._beta_y = self._d0 - self._sy

    # sample a state trajectory from _particles
    def _get_state_trajectory(self):
        if self._particles is None:
            print("call generateWeightedParticles first")
            exit(0)

        x_star = np.zeros(self._seq_len)
        indx = multinomial_resampling(self._normalised_weights[:, self._seq_len - 1], 1)[0]

        for t in reversed(range(self._seq_len)):
            indx = self._ancestors[indx, t]
            x_star[t] = self._particles[indx, t]
        return x_star

    def _generate_particles(self, y, x0, x_ref):
        # Stop, if input parameters are NULL
        if y is None:
            sys.exit("Error: the input parameter is NULL")

        # Number of states
        self._seq_len = len(y)

        # initialize
        self._seq_len = len(y)
        self._particles = np.zeros((self._n_particles, self._seq_len))
        self._normalised_weights = np.zeros((self._n_particles, self._seq_len))
        self._ancestors = np.zeros((self._n_particles, self._seq_len), dtype=int)
        self._sx = np.zeros(self._n_particles)
        self._sy = np.zeros(self._n_particles)
        self._beta_x = np.zeros(self._n_particles)
        self._beta_y = np.zeros(self._n_particles)

        # Init state, at t=0
        self._particles[:, 0] = x0  # Deterministic initial condition

        # Hyperarameters
        self._alpha_x = self._a0
        self._beta_x = self._b0
        self._alpha_y = self._c0
        self._beta_y = self._d0

        # weighting step
        self._weighting_step(y[0], 0)

        for t in range(1, self._seq_len):
            # resampling step
            self._resample_step(t - 1)
            self._ancestors[self._n_particles - 1, t-1] = self._n_particles - 1

            # propagation step
            self._propagation_step(t)
            self._particles[self._n_particles - 1, t] = x_ref[t]

            # weighting step
            self._weighting_step(y[t], t)

        self._ancestors[:, self._seq_len-1] = np.arange(self._n_particles)

    def sample_states(self, y, x0, x_ref, n_particles=100, max_iter=100):

        self._n_particles = n_particles

        x_new_ref = x_ref.copy()

        for i in range(max_iter):
            self._generate_particles(y, x0, x_new_ref)
            x_new_ref = self._get_state_trajectory()

        return x_new_ref


if __name__ == '__main__':
    import matplotlib.pyplot as plt
    from utils import state_transition_func, transfer_func, generate_data
    # Set up some parameters
    n_particle = 100  # Number of particles
    seq_len = 100  # Length of data record
    f1 = state_transition_func
    g1 = transfer_func
    # R = 1.
    # Q = 0.1
    xref = np.zeros(seq_len)
    x_0 = np.zeros(n_particle)
    # Generate data
    data_x, data_y = generate_data(r=1.0, q=0.1, seq_len=seq_len)

    pf = MargCSMC(f1, g1, np.ones(4))

    x_mult = pf.sample_states(data_y, x0=0, x_ref=xref, n_particles=n_particle, max_iter=100)

    plt.plot(data_x, marker='o', label='Real States', markersize=3)
    plt.plot(x_mult, label='Marg CSMC States')

    plt.legend()
    plt.show()

File: RustCode/src/marg_csmc.rs
use rand::{Rng};
use std::process::exit;
use rand::{SeedableRng};
use rand::rngs::StdRng;
use rand_distr::StudentT;
use statrs::distribution::{StudentsT, Continuous};
use crate::utils;

type FDef = fn(&Vec<f64>, usize) -> Vec<f64>;
type GDef = fn(&Vec<f64>) -> Vec<f64>;
pub struct Priors(pub f64, pub f64, pub f64, pub f64);

// Marginal conditional_SMC class
#[allow(dead_code)]
pub struct MargCSMC {
    _f: FDef,
    _g: GDef,
    //_p: Priors,
    _n_particles: usize,
    _a0: f64,
    _b0: f64,
    _c0: f64,
    _d0: f64,
    _alpha_x: f64,
    _alpha_y: f64,
    _beta_x: Vec<f64>,
    _beta_y: Vec<f64>,
    _sx: Vec<f64>,
    _sy: Vec<f64>,
    _seq_len: usize,
    _particles: Vec<Vec<f64>>,
    _normalised_weights: Vec<f64>,
    _ancestors: Vec<Vec<usize>>,
    _rng: StdRng
}

#[allow(dead_code)]
impl MargCSMC {
    pub fn new(f: FDef, g: GDef, p:Priors) -> Self {
        MargCSMC {
            _f: f,
            _g: g,
            //_p: p,
            _n_particles: 100,
            _a0: p.0,
            _b0: p.1,
            _c0: p.2,
            _d0: p.3,
            _alpha_x: 0.0,
            _alpha_y: 0.0,
            _beta_x: Default::default(),
            _beta_y: Default::default(),
            _sx: Default::default(),
            _sy: Default::default(),
            _seq_len: 0,
            _particles: Default::default(),
            _normalised_weights: Default::default(),
            _ancestors: Default::default(),
            _rng: StdRng::from_entropy()
        }
    }

    // sample state trajectory from a set of (weighted) particles
    pub fn _get_state_trajectory(&mut self) -> Vec<f64> {
        if self._seq_len == 0 {
            println!("call generateWeightedParticles first");
            exit(0);
        }

        let mut x_star: Vec<f64> = vec![0.0; self._seq_len as usize]; //Default::default();
        let mut indx: usize = utils::multinomial_resampling(&mut self._rng,
                                                             &self._normalised_weights,
                                                             &1_usize)[0];
        for t in (0..self._seq_len).rev() {
            indx = self._ancestors[t][indx];
            x_star[t] = self._particles[t][indx];
        }
        x_star
    }

    // computes weights of particles set at a given time t, and observation y
    fn _weighting_step(&mut self, y_t:f64, t: usize) {
        let mut logweights: Vec<f64> = vec![0.0; self._n_particles];
        let mut max_weight: f64;
        let nu = 2.0 * self._alpha_y;
        let particles = & self._particles[t];
        let ypred:Vec<f64> = (self._g)(particles);

        for i in 0..self._n_particles {
            let sig2 = self._beta_y[i] / self._alpha_y;
            let t_dist = StudentsT::new(y_t, sig2.sqrt(), nu).unwrap();
            logweights[i] = t_dist.pdf(ypred[i]).ln();
        }

        max_weight = logweights[0];
        for i in 0..self._n_particles
        {
            if max_weight < logweights[i] {
                max_weight = logweights[i];
            }
        }
        // log sum trick to avoid numerical underflow
        let weights:Vec<f64> = logweights.iter().map(|x| (x-max_weight).exp()).collect();
        let w_sum:f64 = weights.iter().sum();
        // compute normalized weights
        self._normalised_weights = weights.iter().map(|x| x /w_sum).collect();
    }

    // update statistics and parameters
    fn _update_stat_param(&mut self, y:&Vec<f64>, t: usize) {
        let particles = & self._particles[t];
        let ypred:Vec<f64> = (self._g)(particles);

        // update hyper params
        self._alpha_x = self._alpha_x + 0.5;
        self._alpha_y = self._alpha_y + 0.5;

        if t>0 {
            let anc:& mut Vec<usize> = & mut self._ancestors[t-1];
            let particles = & self._particles[t-1];
            let xpred:Vec<f64> = (self._f)(particles, t - 1);
            let cur_particles = & self._particles[t];

            for n in 0..self._n_particles {
                let indx_n: usize = anc[n];
                self._sx[n] = self._sx[indx_n] - 0.5 * (cur_particles[n] - xpred[indx_n]).powi(2);
                self._sy[n] = self._sy[indx_n] - 0.5 * (y[t] - ypred[n]).powi(2);
                self._beta_x[n] = self._b0 - self._sx[n];
                self._beta_y[n] = self._d0 - self._sy[n];
            }
        }
        else {
            for n in 0..self._n_particles {
                self._sy[n] =  -0.5 * (y[t] - ypred[n]).powi(2);
                self._beta_x[n] = self._b0;
                self._beta_y[n] = self._d0 - self._sy[n];
            }
        }
    }

    fn _resample_step(&mut self, t:usize){
        let mut anc = utils::multinomial_resampling(&mut self._rng,
                                                              &self._normalised_weights,
                                                              &0_usize);
        anc[self._n_particles-1] = self._n_particles-1;
        self._ancestors[t] = anc;
    }

    fn _propagation_step(&mut self, x_ref:f64, t:usize){
        let nu = 2.0 * self._alpha_x;
        let t_dist = StudentT::new(nu).unwrap();
        let particles = & self._particles[t-1];
        let xpred:Vec<f64> = (self._f)(particles, t - 1);
        let cur_particles: & mut Vec<f64> =  & mut self._particles[t];

        for n in 0..(self._n_particles-1) {
            let indx_n:usize = self._ancestors[t-1][n];
            let sig2 = self._beta_x[indx_n] / self._alpha_x;
            let rand_t:f64 = self._rng.sample(t_dist);

            cur_particles[n] = xpred[n] + sig2.sqrt() * rand_t;
        }
        cur_particles[self._n_particles-1] = x_ref;
    }
    // generate particles
    fn _generate_particles(&mut self, y: &Vec<f64>, x0: f64, x_ref: &Vec<f64>) {
        // Initialization of variables
        self._alpha_x = self._a0;
        self._alpha_y = self._c0;
        self._beta_x = vec![self._b0; self._n_particles];
        self._beta_y = vec![self._d0; self._n_particles];
        self._sx = vec![0.0; self._n_particles];
        self._sy = vec![0.0; self._n_particles];

        self._seq_len = y.len(); // Number of states
        self._particles = vec![vec![0.0; self._n_particles]; self._seq_len];
        self._ancestors = vec![vec![0; self._n_particles]; self._seq_len];
        self._normalised_weights = vec![0.0; self._n_particles];

        // deterministically set the initial state
        self._particles[0].iter_mut().for_each(|x| *x = x0);

        //weighting step at t=0
        self._weighting_step(y[0], 0);
        self._update_stat_param(y, 0);

        for t in 1..self._seq_len {
            //resampling step
            //println!("{}", t);
            self._resample_step(t-1);

            // propagation step
            self._propagation_step(x_ref[t], t);

            // weighting step
            self._weighting_step(y[t], t);
            self._update_stat_param(y, t);
        }
        //set the last column of ancestors as [0,1,2,...N-1]
        self._ancestors[self._seq_len - 1] = (0..self._n_particles).collect();
    }

    // generate particles
    pub fn sample_states(&mut self, y: &Vec<f64>, x0: f64, x_ref: &Vec<f64>, n_particles: usize,
                         max_iter:usize)->Vec<f64> {
        let mut x_ref_new:Vec<f64> = x_ref.clone();
        self._n_particles = n_particles;
        for _i in 0..max_iter{
            self._generate_particles(y, x0, &x_ref_new);
            x_ref_new = self._get_state_trajectory();
        }
        x_ref_new
    }
}